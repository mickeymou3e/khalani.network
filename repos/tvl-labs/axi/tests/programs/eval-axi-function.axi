function factorial(n: Nat) -> Nat {
    match n {
        Zero => 1,
        Succ(m) => n mul factorial(m),
    }
}

!prove_by_eval_term(factorial(0));
!prove_by_eval_term(factorial(1));
!prove_by_eval_term(factorial(5));

function fibonacci(n: Nat) -> Nat {
    match n {
        Succ(Zero) => 1,
        Succ(Succ(m)) => fibonacci(m) add fibonacci(m add 1),
        Zero => 0,
    }
}

!prove_by_eval(fibonacci(0) = 0);
!prove_by_eval(fibonacci(1) = 1);
!prove_by_eval(fibonacci(8) = 21);

inductive ListNat {
    ListNatNil,
    ListNatCons(Nat, ListNat),
}

function sum_list_nat(xs: ListNat) -> Nat {
    let ys := xs;
    let zero := Zero;

    match ys {
        ListNatNil => zero,
        ListNatCons(x, xs) => x add sum_list_nat(xs),
    }
}

!prove_by_eval(sum_list_nat(
    ListNatCons(
        1,
        ListNatCons(
            2,
            ListNatCons(
                3,
                ListNatNil,
            ),
        ),
    )
) = 6);

function my_predicate(x: Nat) -> Boolean {
    x lt 3 | x = 3
}

!prove_by_eval(my_predicate(1));
!prove_by_eval(my_predicate(3));
!prove_by_eval(my_predicate(Succ(Succ(Succ(Zero)))));

function my_boolean_function() -> Boolean {
    my_predicate(1) | false
}

!prove_by_eval(my_boolean_function());

# Should fail.
!prove_by_eval(my_predicate(4));
