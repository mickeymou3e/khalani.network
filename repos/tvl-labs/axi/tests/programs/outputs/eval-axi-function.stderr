[31mError:[0m 
    [38;5;246mâ•­[0m[38;5;246mâ”€[0m[38;5;246m[[0m<unknown>:67:16[38;5;246m][0m
    [38;5;246mâ”‚[0m
 [38;5;246m67 â”‚[0m [38;5;249m![0m[38;5;249mp[0m[38;5;249mr[0m[38;5;249mo[0m[38;5;249mv[0m[38;5;249me[0m[38;5;249m_[0m[38;5;249mb[0m[38;5;249my[0m[38;5;249m_[0m[38;5;249me[0m[38;5;249mv[0m[38;5;249ma[0m[38;5;249ml[0m[38;5;249m([0mmy_predicate(4)[38;5;249m)[0m[38;5;249m;[0m
 [38;5;240m   â”‚[0m                â”€â”€â”€â”€â”€â”€â”€â”¬â”€â”€â”€â”€â”€â”€â”€  
 [38;5;240m   â”‚[0m                       â•°â”€â”€â”€â”€â”€â”€â”€â”€â”€ prove_by_eval failed because the result is not true
[38;5;246mâ”€â”€â”€â”€â•¯[0m
